taskcluster_util/util/downloader.py

```diff
@@ -54,23 +54,35 @@ def download_latest_artifact(self, task_id, full_filename, dest_dir):
         sys.stdout.flush()
 
         # move file to dest folder
-        abs_dest_dir = os.path.abspath(dest_dir)
-        log.debug('dest dir: {}'.format(abs_dest_dir))
-        if os.path.exists(abs_dest_dir) and (not os.path.isdir(abs_dest_dir)):
-            log.warning('Not a directory: {}'.format(abs_dest_dir))
-            final_file_path = os.path.abspath(temp_local_file)
-            log.debug('local file: {}'.format(final_file_path))
-        else:
-            if not os.path.exists(abs_dest_dir):
-                os.makedirs(abs_dest_dir)
-            shutil.copy(temp_local_file, abs_dest_dir)
-            final_file_path = os.path.join(abs_dest_dir, base_filename)
-            log.debug('local file: {}'.format(final_file_path))
-
-        # remove temp folder
+        final_file_path = temp_local_file
+        abs_dest_dir = os.path.abspath(dest_dir) if dest_dir else os.getcwd()
         try:
-            shutil.rmtree(self.temp_dir)  # delete directory
-        except OSError:
-            log.warning('Can not remove temporary folder: {}'.format(self.temp_dir))
+            FolderHandler(abs_dest_dir).copy_elements_from(temp_local_file)
+            final_file_path = os.path.join(abs_dest_dir, base_filename)
+            # remove temp folder
+            try:
+                shutil.rmtree(self.temp_dir)  # delete directory
+            except OSError:
+                logger.warning('Can not remove temporary folder: {}'.format(self.temp_dir))
+        except Exception as e:
+            logger.error(e.message)
+            logger.debug('local file: [{}]'.format(temp_local_file))
 
         return final_file_path
+
+
+class FolderHandler:
+    def __init__(self, path):
+        self.path = path
+        self._check_if_folder_is_valid()
+
+    def _check_if_folder_is_valid(self):
+        if not os.path.exists(self.path):
+            raise Exception("[{}] doesn't exist.".format(self.path))
```

Comment:
After taking a look at what wget does, they do create a new folder if it doesn't already exist. We can extract the try logic in a function, so `_check_if_folder_is_valid()` keeps only 1 responsibility (checking the state of the folder). Also, the message in the exception can be replaced by 'Can not create the folder', because that's what we exactly failed to do.

```python
def _check_if_folder_is_valid(self):
    if not os.path.exists(self.path):
        logger.debug('[{}] doesn't exist, trying to create it')
        self.create()
    if not os.path.isdir(self.path):
        ...

def create(self):
    try:
        os.makedirs(self.path)
    except os.error as e:
        logger.debug('errno: [{}], strerror: [{}], filename: [{}]'.format(e.errno, e.strerror, e.filename))
        raise Exception('Can not create the folder: [{}]'.format(self.path))
```